src/protocolsupport/utils/netty/Decompressor.java

```diff
@@ -19,33 +24,55 @@ public static Decompressor create() {
 		return recycler.get();
 	}
 
-	private final Inflater inflater = new Inflater();
+	private final BungeeZlib zlib = CompressFactory.zlib.newInstance();
 	private final Handle<Decompressor> handle;
+
 	protected Decompressor(Handle<Decompressor> handle) {
 		this.handle = handle;
+		zlib.init(false, Deflater.DEFAULT_COMPRESSION);
 	}
 
-	public byte[] decompress(byte[] input, int uncompressedlength) throws DataFormatException {
-		byte[] uncompressed = new byte[uncompressedlength];
+	public void decompress(ByteBuf in, ByteBuf out) {
+		ByteBuf tmp = null;
+		if (!in.hasMemoryAddress()) {
+			tmp = in.alloc().directBuffer(in.readableBytes());
+			tmp.writeBytes(in);
+		}
+		if (tmp != null) {
+			tmp.memoryAddress();
+		}
+		out.memoryAddress();
 		try {
-			inflater.setInput(input);
-			inflater.inflate(uncompressed);
+			zlib.process(tmp == null ? in : tmp, out);
+		} catch (DataFormatException e) {
+			throw new RuntimeException("Failed decompressing packet", e);
 		} finally {
-			inflater.reset();
+			ReferenceCountUtil.release(tmp);
 		}
-		return uncompressed;
 	}
 
 	public void recycle() {
 		handle.recycle(this);
 	}
 
-	public static byte[] decompressStatic(byte[] input, int uncompressedlength) throws DataFormatException {
-		Decompressor decompressor = create();
+	@Override
+	public void finalize() throws Throwable {
+		zlib.free();
```

Comment:
That doesn't work. The zlib object itself can gced before wrapper object.

Reply:
i think im going to revert the zlib stuff in this branch. it got accidentally copied over from my dev branch